File: daemon/handlers/packet_dispatcher.cpp
#include "packet_dispatcher.hpp"
#include "discovery_handler.hpp"
#include "decision_handler.hpp"
#include "correction_handler.hpp"
#include "observation_handler.hpp"
#include "../../common/logger.hpp"

PacketDispatcher::PacketDispatcher() {}

bool PacketDispatcher::parse_header(const std::unordered_map<std::string, std::string>& packet, PacketHeader& header) {
    // Required fields: message_id, type, source, destination, sequence
    try {
        header.message_id = packet.at("message_id");
        header.type = packet.at("type");
        header.source = packet.at("source");
        header.destination = packet.at("destination");
        header.sequence = std::stoi(packet.at("sequence"));
        return true;
    } catch (std::exception& e) {
        Logger::error("PacketDispatcher: header parsing error: " + std::string(e.what()));
        return false;
    }
}

bool PacketDispatcher::dispatch(const std::unordered_map<std::string, std::string>& packet) {
    PacketHeader header;
    if (!parse_header(packet, header)) {
        Logger::error("PacketDispatcher: invalid/corrupt packet dispatch aborted.");
        return false;
    }

    Logger::info("PacketDispatcher: dispatching packet type: " + header.type + " (id=" + header.message_id + ")");
    try {
        route_packet(header, packet);
    } catch (std::exception& exc) {
        Logger::error("PacketDispatcher: dispatch error for type '" + header.type + "': " + exc.what());
        return false;
    }
    return true;
}

void PacketDispatcher::route_packet(const PacketHeader& header, const std::unordered_map<std::string, std::string>& packet) {
    std::string type = header.type;
    std::transform(type.begin(), type.end(), type.begin(), ::tolower);
    if (type == "discovery") {
        handle_discovery_packet(packet);
    } else if (type == "decision") {
        handle_decision_packet(packet);
    } else if (type == "correction") {
        handle_correction_packet(packet);
    } else if (type == "observation" || type == "telemetry") {
        handle_observation_packet(packet);
    } else {
        Logger::warning("PacketDispatcher: Unknown message type '" + type + "'");
        throw std::runtime_error("Unknown message type");
    }
}

File: daemon/handlers/packet_dispatcher.py
"""
ATROP Protocol Packet Dispatcher

Routes packets to appropriate handler based on message type.
Handles:
 - Discovery
 - Decision
 - Correction
 - Observation/Telemetry

Parses and validates basic ATROP headers:
 - message_id, type, source, destination, sequence
Logs errors and supports both simulated/test source (raw dict/bytes) and real inputs.

Usage:
    dispatcher = PacketDispatcher()
    dispatcher.dispatch(raw_packet)
"""

from logger import setup_logger
from handlers.discovery_handler import handle_discovery_packet
from handlers.decision_handler import handle_decision_packet
from handlers.correction_handler import handle_correction_packet
from handlers.observation_handler import handle_observation_packet

class PacketDispatcher:
    def __init__(self, logger_name="ATROP.Dispatcher", log_config=None):
        self.log = setup_logger(logger_name, log_config or {})

    def parse_header(self, raw_packet):
        """
        Expecting raw_packet as dict for simulation. In prod, decode bytes here.
        Expected keys: message_id, type, source, destination, sequence
        """
        try:
            header = {
                'message_id': raw_packet['message_id'],
                'type': raw_packet['type'],
                'source': raw_packet.get('source', ''),
                'destination': raw_packet.get('destination', ''),
                'sequence': raw_packet.get('sequence', 0),
            }
            return header
        except (KeyError, TypeError) as e:
            self.log.error(f"Header parsing error: {e}")
            return None

    def dispatch(self, raw_packet):
        header = self.parse_header(raw_packet)
        if not header:
            self.log.error("Invalid or corrupt packet; dispatch aborted.")
            return False

        msg_type = header['type'].lower()
        self.log.info(f"Dispatching packet type: {msg_type} (id={header['message_id']})")

        try:
            if msg_type == "discovery":
                handle_discovery_packet(raw_packet)
            elif msg_type == "decision":
                handle_decision_packet(raw_packet)
            elif msg_type == "correction":
                handle_correction_packet(raw_packet)
            elif msg_type in ("observation", "telemetry"):
                handle_observation_packet(raw_packet)
            else:
                self.log.warning(f"Unknown message type '{msg_type}'")
                return False
        except Exception as exc:
            self.log.error(f"Dispatch error for type '{msg_type}': {exc}")
            return False

        return True
